presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,222 @@
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverValueCount;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * This operator acts as a simple "pass-through" pipe, while saving its input pages.
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterCollectOperator
+        implements Operator
+{
+    public static class DynamicFilterCollectOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final Consumer<TupleDomain<Integer>> dynamicPredicateConsumer;
+        private final List<Integer> channels;
+        private final List<Type> buildTypes;
+
+        public DynamicFilterCollectOperatorFactory(int operatorId, PlanNodeId planNodeId, Consumer<TupleDomain<Integer>> dynamicPredicateConsumer, List<Integer> channels, List<Type> buildTypes)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicPredicateConsumer = requireNonNull(dynamicPredicateConsumer, "dynamicPredicateConsumer is null");
+            this.channels = requireNonNull(channels, "channels is null");
+            verify(channels.stream().collect(toSet()).size() == channels.size(), "duplicate channels are not allowed");
+            this.buildTypes = requireNonNull(buildTypes, "buildTypes is null");
+            verify(buildTypes.size() == channels.size(), "channels and buildTypes must have same length");
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            return new DynamicFilterCollectOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterCollectOperator.class.getSimpleName()),
+                    dynamicPredicateConsumer,
+                    channels,
+                    buildTypes,
+                    planNodeId);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("duplicate() is not supported for DynamicFilterCollectOperatorFactory");
+        }
+    }
+
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final Consumer<TupleDomain<Integer>> dynamicPredicateConsumer;
```

Comment:
Could you pass dynamic filter string id as tuple domain instead?

Channel index is specific to local execution and might be harder to interpret when we send dynamic filter to coordinator

Reply:
Sounds good, refactored `LocalDynamicFilter` a bit to support that.